xwiki-platform-core/xwiki-platform-wysiwyg/xwiki-platform-wysiwyg-api/src/main/java/org/xwiki/wysiwyg/internal/converter/DefaultRequestParameterConverter.java

```diff
@@ -0,0 +1,223 @@
+     * that required HTML conversion.
+     */
+    private static final String CONVERSION_OUTPUT = "com.xpn.xwiki.wysiwyg.server.converter.output";
+
+    /**
+     * The name of the session attribute holding the conversion exceptions. The conversion exceptions are stored in a
+     * {@link Map} of {@link Map}s. The first key identifies the request and the second key is the name of the request
+     * parameter that required HTML conversion.
+     */
+    private static final String CONVERSION_ERRORS = "com.xpn.xwiki.wysiwyg.server.converter.errors";
+
+    @Inject
+    private MutableServletRequestFactory mutableServletRequestFactory;
+
+    @Inject
+    private HTMLConverter htmlConverter;
+
+    @Inject
+    @Named("context")
+    private ComponentManager componentManager;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public Optional<ServletRequest> convert(ServletRequest request, ServletResponse response) throws IOException
+    {
+        Optional<ServletRequest> result = Optional.of(request);
+        // Take the list of request parameters that require HTML conversion.
+        String[] parametersRequiringHTMLConversion = request.getParameterValues(REQUIRES_HTML_CONVERSION);
+        if (parametersRequiringHTMLConversion != null) {
+            // Wrap the current request in order to be able to change request parameters.
+            MutableServletRequest mreq = this.mutableServletRequestFactory.newInstance(request);
+            // Remove the list of request parameters that require HTML conversion to avoid recurrency.
+            mreq.removeParameter(REQUIRES_HTML_CONVERSION);
+            // Try to convert each parameter from the list and save caught exceptions.
+            Map<String, Throwable> errors = new HashMap<String, Throwable>();
+            // Save also the output to prevent loosing data in case of conversion exceptions.
+            Map<String, String> output = new HashMap<String, String>();
+            convertHTMLWithExecutionContext(parametersRequiringHTMLConversion, mreq, output, errors);
+            if (!errors.isEmpty()) {
+                handleConversionErrors(errors, output, mreq, response);
+                result = Optional.empty();
+            } else {
+                result = Optional.of(mreq);
+            }
+        }
+        return result;
+    }
+
+    private void convertHTMLWithExecutionContext(String[] parametersRequiringHTMLConversion,
+        MutableServletRequest request, Map<String, String> output, Map<String, Throwable> errors)
+    {
+        new ExecutionContextRunnable(() -> convertHTML(parametersRequiringHTMLConversion, request, output, errors),
+            this.componentManager).run();
+    }
+
+    private void convertHTML(String[] parametersRequiringHTMLConversion, MutableServletRequest request,
+        Map<String, String> output, Map<String, Throwable> errors)
+    {
+        for (int i = 0; i < parametersRequiringHTMLConversion.length; i++) {
```

Comment:
can be slightly simplified using `for (String parameterName : parametersRequiringHTMLConversion) {`